File: 0002-add-two-numbers/simple.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = tail = ListNode()
        carry = 0

        while ((l1 is not None) or (l2 is not None) or (carry != 0)):
            sum = 0

            if (carry != 0):
                sum = sum + 1
                carry = carry - 1
            
            if l1:
                sum += l1.val
                l1 = l1.next

            if l2:
                sum += l2.val
                l2 = l2.next

            if (sum >= 10):
                carry = carry + 1
                sum = sum % 10
            
            tail.next = ListNode(sum)
            tail = tail.next

        return head.next
        

File: 0002-add-two-numbers/simple.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

use std::ptr;

impl Solution
{
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>>
    {
        let mut a = &l1;
        let mut b = &l2;

        let mut head = ListNode::new(0);
        let mut tail = &mut head;
        let mut carry = 0;

        while ((a.is_some()) || (b.is_some()) || (carry != 0))
        {
            let mut sum = 0;

            if (carry != 0)
            {
                sum = sum + 1;
                carry = carry - 1;
            }

            if let Some(node) = a
            {
                sum += node.val;
                a = &node.next;
            }

            if let Some(node) = b
            {
                sum += node.val;
                b = &node.next;
            }

            if (sum >= 10)
            {
                carry = carry + 1;
                sum = sum & 10;
            }

            tail.next = ListNode::new(sum);
            tail = tail.next;
        }

        return head.next;

    }
}